crates/bevy_ecs/hecs/src/dynamic_query.rs

```diff
@@ -0,0 +1,527 @@
+    any::TypeId,
+    hash::Hash,
+    mem::{align_of, size_of},
+    ptr::{slice_from_raw_parts, slice_from_raw_parts_mut, NonNull},
+};
+use std::{sync::Mutex, vec::Vec};
+
+use crate::{archetype::ComponentIdSet, Access, Archetype, ComponentId, Entity, TypeInfo};
+
+pub struct DynamicComponentInfoRegistry {
+    /// Collection of dynamic component info that is kept to ensure the consistency of the info for
+    /// dynamically created components.
+    dynamic_components: Arc<Mutex<HashMap<u64, DynamicComponentInfo>>>,
+}
+
+impl DynamicComponentInfoRegistry {
+    pub(crate) fn new() -> Self {
+        Self {
+            dynamic_components: Default::default(),
+        }
+    }
+}
+
+impl DynamicComponentInfoRegistry {
+    /// Get the component info for the given Rust type
+    pub fn get_rust_component_info<T: 'static>(&self) -> DynamicComponentInfo {
+        DynamicComponentInfo {
+            id: TypeId::of::<T>().into(),
+            layout: Layout::from_size_align(size_of::<T>(), align_of::<T>()).unwrap(),
+            drop: None,
+        }
+    }
+
+    /// Registers a dynamic component and returns [`DynamicComponentInfo`] which can be used to
+    /// spawn dynamic components with an [`EntityBuilder`].
+    ///
+    /// Returns `None` if a component with the provided ID has already been registered.
+    pub fn register_dynamic_component(
+        &self,
+        external_id: u64,
+        layout: Layout,
+        drop: Option<fn(*mut u8)>,
+    ) -> Option<DynamicComponentInfo> {
+        let mut dynamic_components = self.dynamic_components.lock().unwrap();
+
+        // If the component has already been registered
+        if dynamic_components.contains_key(&external_id) {
+            // Return none, we can't re-register the same component
+            None
+
+        // If the component has not been registered yet
+        } else {
+            // Register the component and return its info
+            let info = DynamicComponentInfo {
+                id: ComponentId::ExternalId(external_id),
+                layout,
+                drop,
+            };
+            dynamic_components.insert(external_id, info);
+
+            Some(info)
+        }
+    }
+
+    /// Return the component info for a non-Rust dynamic component. If the dynamic component has not
+    /// been registered yet, it will return None.
+    pub fn get_dynamic_component_info(&self, external_id: u64) -> Option<DynamicComponentInfo> {
+        self.dynamic_components
+            .lock()
+            .unwrap()
+            .get(&external_id)
+            .map(Clone::clone)
+    }
+}
+
+lazy_static::lazy_static! {
+    /// The global registry of component info
+    pub static ref DYNAMIC_COMPONENT_INFO_REGISTRY: DynamicComponentInfoRegistry =
+        DynamicComponentInfoRegistry::new();
+}
+
+/// A Query that can be constructed at runtime
+#[derive(Default)]
+pub struct DynamicQuery {
+    /// Whether or not the entity should be queried
+    pub entity: bool,
```

Comment:
Is the complexity of making this optional justified?

Reply:
I don't know. It doesn't seem to be a big deal to me to have the flag, but it probably isn't a big deal just to include the entity all the time is it? The only reason not to would be the extra overhead, but it's probably tiny overhead anyway.

Reply:
I'm inclined to prioritize simplicity if there's not a specific requirement otherwise. With careful inlining an unused value will be optimized out anyway; this is observable in practice for static queries.

Reply:
Sweet, sounds good.